chain/chain/src/chain.rs

```diff
@@ -1221,6 +1217,45 @@ impl Chain {
     }
 
     fn process_block_single<F, F1, F2, F3>(
+        &mut self,
+        me: &Option<AccountId>,
+        block: MaybeValidated<Block>,
+        provenance: Provenance,
+        block_accepted: F,
+        block_misses_chunks: F1,
+        orphan_misses_chunks: F2,
+        on_challenge: F3,
+    ) -> Result<Option<Tip>, Error>
+    where
+        F: FnMut(AcceptedBlock),
+        F1: Copy + FnMut(BlockMissingChunks),
+        F2: Copy + FnMut(OrphanMissingChunks),
+        F3: FnMut(ChallengeBody),
+    {
+        metrics::BLOCK_PROCESSING_ATTEMPTS_TOTAL.inc();
+        metrics::NUM_ORPHANS.set(self.orphans.len() as i64);
+        let success_timer = metrics::BLOCK_PROCESSING_TIME.start_timer();
+
+        let res = self.process_block_single_impl(
+            me,
+            block,
+            provenance,
+            block_accepted,
+            block_misses_chunks,
+            orphan_misses_chunks,
+            on_challenge,
+        );
+
+        if res.is_ok() {
+            metrics::BLOCK_PROCESSED_TOTAL.inc();
+            success_timer.stop_and_record();
+        } else {
+            success_timer.stop_and_discard();
+        }
+        res
+    }
+
+    fn process_block_single_impl<F, F1, F2, F3>(
```

Comment:
please add comments on what this function does (and how does it differ from 'process_block_single')